priv/posts/performance-methodologies.md

```diff
@@ -0,0 +1,170 @@
+
+Logging (include system log) can be categorized as low-frequency tracing that is enabled by default.
+
+Some common tools:
+
+**system-wide**:
+- `tcpdump`: network packet tracing
+- `perf`: Linux Performance Events (tracing static and dynamic probes)
+
+**per-process**:
+- `strace`: system call tracing
+- `USDT` (Userland Statically Defined Tracing)
+- `DTrace`: observability framework that includes a programming language and a tool.
+
+TracePoints is a great way to observe your software in the production environment. You can use USDT (dynamic probes) or static tracepoints.
+For further information check the _useful links_ section.
+
+### Profiling
+
+Profiling characterizes the target by collecting a set of samples of snapshots. CPU usage is a common example where samples are taken of the stack trace to characterize the code paths that are consuming CPU cycles.
+
+**Note**: For further information about Profiling CPU, I've made a blog post doing CPU Profiling in a Node.js application. [Check here](https://blog.rafaelgss.com.br/node-cpu-profiler).
+
+Tools:
+
+- `perf`: Linux Performance Events (profiling)
+- `cachegrind`: a Valgrind sub tool, can profile hardware cache usage and be visualized using `kcachegrind`
+
+> `/proc` is a file system interface for kernel statistics, it contains directories where each directory is named after the **PID** of the process. These directories contain a number of files containing information and statistics about each process mapped from kernel data structures.
+
+![Julia Evans - Comic /proc](https://pbs.twimg.com/media/DZ3HpVXXkAEgxpc?format=jpg&name=large) - [reference](https://twitter.com/b0rk/status/981159808832286720/photo/1)
+
+## Methodologies
+
+This section will describe three of most used methodologies. Apply a methodology due performance issues is needed, and there is no rule to choose the best approach,
+I would say that past experiences and knowledge in the software definitely will help with that.
+
+### USE
+
+Utilization, Saturation and Errors (USE) is an methodology that **should be used early in performance investigation**. For every resource, check the utilization, saturation, and errors:
+
+- **Resource**: all physical server functional components (CPU, busses, ...)
+- **Utilization**: for a set time interval, the percentage of time that the resource was busy servicing work. While busy, the resource may **still be able to accept more work**.
+- **Saturation**: Extra work, often waiting on a queue. Are jobs that service cannot deal at moment.
+
+![Workflow with USE Methodology](/images/performance-analysis/workflow-use.png)
+
+However, this could be counter-intuitive, a short burst of high utilization can cause saturation and performance issues, even though the overall utilization is low over a long interval. CPU utilization, for example, **can vary dramatically from second to second, so a 5-minute average may disguise short periods of 100% utilization and therefore saturation**.
```

Comment:
```suggestion
Its important to consider that it can be counter-intuitive; a short burst of high utilization can introduce saturation and performance issues even though the overall utilization is low over a long interval. CPU utilization **can change dramatically from second to second** so a 5-minute average may disguise short periods of 100% utilization and therefore lead to saturation.
```